replicatedhq/ship
# mechanism for setting github contents during local dev

What I Did
------------

Add a flag for overriding github repo contents when running `ship app`
with the `--runbook` flag for local dev.

We allow for the user to pass a flag `--set-github-contents`, and parse
a string of the form

    REPO:REPO_PATH:REF:LOCAL_PATH

e.g.

    replicatedhq/test-charts:/plain-k8s:master:$GOPATH/src/github.com/replicatedhq/test-charts

This will override an asset of the form

```yaml
assets:
  v1:
  - github:
      repo: replicatedhq/test-charts
      path: /plain-k8s
      ref: master
      # ...
```

with the contents of the local file system.

This allows for much faster iteration on ship-vendor apps that use the `github` asset type. Up to this point, such apps **had** to be built, pulled, and tested via console.replicated.com. Now you can "fake" the github assets by dropping in local paths to the checked-out repos.

I know `ship app` is going away eventually, but most of this can be reused to provide the same functionality for `ship init file::path/to/ship.yaml`.

How I Did it
------------

- Add a flag for `--set-github-contents`
- If `--runbook` is passed, loop over `set-github-contents` values, and override repository contents in the `api.Release#GithubContent` by walking the local filesystem.

Still TODO integration tests

How to verify it
------------

create at /tmp/ship/ship.yaml

```yaml
lifecycle:
  v1:
    - config: {}
    - render:
        root: ./
assets:
  v1:
  - github:
      repo: replicatedhq/test-charts
      path: /base
      ref: master
      dest: repo
```

create at /tmp/repo/base/deployment.yaml

```
fake
```

run ship with the flag

```
ship app --runbook /tmp/ship/ship.yaml --set-github-contents replicatedhq/test-charts:/base:master:/tmp/repo
```

Verify that one of the output files is

```
./base/deployment.yaml
```

with contents

```
fake
```

Description for the Changelog
------------

Add `--set-github-contents` flag for overriding [github assets](https://ship.replicated.com/reference/assets/github/) during local development of a `ship.yaml`.


What I Did
------------


How I Did it
------------


How to verify it
------------


Description for the Changelog
------------



Picture of a Boat (not required but encouraged)
------------
